contracts/FastBreak.cdc

```diff
@@ -0,0 +1,718 @@
+            fastBreakGameID: String,
+            topShots: [UInt64]
+        ): @FastBreak.NFT
+    }
+
+    /// Fast Break game collection
+    ///
+    pub resource Collection:
+        NonFungibleToken.Provider,
+        NonFungibleToken.Receiver,
+        NonFungibleToken.CollectionPublic,
+        FastBreakNFTCollectionPublic,
+        FastBreakPlayer
+    {
+
+        pub var ownedNFTs: @{UInt64: NonFungibleToken.NFT}
+        pub var numMinted: UInt64
+
+        pub fun withdraw(withdrawID: UInt64): @NonFungibleToken.NFT {
+            let token <- self.ownedNFTs.remove(key: withdrawID) ?? panic("Could not find a fast break with the given ID in the Fast Break collection")
+
+            emit Withdraw(id: token.id, from: self.owner?.address)
+
+            return <-token
+        }
+
+        pub fun deposit(token: @NonFungibleToken.NFT) {
+            let token <- token as! @FastBreak.NFT
+            let id: UInt64 = token.id
+
+            let oldToken <- self.ownedNFTs[id] <- token
+
+            emit Deposit(id: id, to: self.owner?.address)
+
+            destroy oldToken
+        }
+
+        pub fun batchDeposit(tokens: @NonFungibleToken.Collection) {
+            let keys = tokens.getIDs()
+
+            for key in keys {
+                self.deposit(token: <-tokens.withdraw(withdrawID: key))
+            }
+
+            destroy tokens
+        }
+
+        pub fun getIDs(): [UInt64] {
+            return self.ownedNFTs.keys
+        }
+
+        pub fun borrowNFT(id: UInt64): &NonFungibleToken.NFT {
+            return (&self.ownedNFTs[id] as &NonFungibleToken.NFT?)!
+        }
+
+        pub fun borrowNFTSafe(id: UInt64): &NonFungibleToken.NFT? {
+            return (&self.ownedNFTs[id] as &NonFungibleToken.NFT?)
+        }
+
+        pub fun borrowFastBreakNFT(id: UInt64): &FastBreak.NFT? {
+            if self.ownedNFTs[id] != nil {
+                let ref = (&self.ownedNFTs[id] as auth &NonFungibleToken.NFT?)!
+                return ref as! &FastBreak.NFT
+            } else {
+                return nil
+            }
+        }
+
+        /// Play the game of Fast Break with an array of Top Shots
+        /// Each account must own a top shot collection to play fast break
+        ///
+        pub fun play(
+            fastBreakGameID: String,
+            topShots: [UInt64]
+        ): @FastBreak.NFT {
+            pre {
+                FastBreak.fastBreakGameByID.containsKey(fastBreakGameID): "no such fast break game"
+                FastBreak.fastBreakGameByID[fastBreakGameID]!.isPublic: "fast break game is private"
+            }
+
+            let acct = getAccount(self.owner?.address!)
+            let collectionRef = acct.getCapability(/public/MomentCollection)
+                                    .borrow<&{TopShot.MomentCollectionPublic}>()!
+
+            /// Must own Top Shots to play Fast Break
+            for flowId in topShots {
+                if !collectionRef.getIDs().contains(flowId) {
```

Comment:
`getIDs()` gets all of the nft keys as an array from the collection, I wonder if there could be issue iterating over an array the length of the users collection `O(topShots*collectionSize)` times.

One suggestion is attempt to borrow each nft `collectionRef.borrowNFT(id: flowId)` which is O(1) map lookup

Reply:
Good call out. I thought it was a map for some reason. Updated to your suggestion for the reasons you mentioned. Thank you.